src/main/java/gregtech/api/capability/IMultipleTankHandler.java

```diff
@@ -1,21 +1,210 @@
 
+    /**
+     * @return Number of tanks in this tank handler
+     */
     int getTanks();
 
-    IFluidTank getTankAt(int index);
-
-    int getIndexOfFluid(FluidStack other);
+    @Nonnull
+    MultiFluidTankEntry getTankAt(int index);
 
+    /**
+     * @return {@code false} if insertion to this fluid handler enforces input to be
+     * filled in one slot at max. {@code true} if it bypasses the rule.
+     */
     boolean allowSameFluidFill();
+
+    /**
+     * Tries to search tank with contents equal to {@code fluidStack}. If {@code fluidStack} is
+     * {@code null}, an empty tank is searched instead.
+     *
+     * @param fluidStack Fluid stack to search index
+     * @return Index corresponding to tank at {@link #getFluidTanks()} with matching
+     */
+    default int getIndexOfFluid(@Nullable FluidStack fluidStack) {
+        List<MultiFluidTankEntry> fluidTanks = getFluidTanks();
+        for (int i = 0; i < fluidTanks.size(); i++) {
+            FluidStack tankStack = fluidTanks.get(i).getFluid();
+            if (fluidStack == tankStack || tankStack != null && tankStack.isFluidEqual(fluidStack)) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    @Override
+    default Iterator<MultiFluidTankEntry> iterator() {
+        return getFluidTanks().iterator();
+    }
+
+    /**
+     * Entry of multi fluid tanks. Retains reference to original {@link IMultipleTankHandler} for accessing
+     * information such as {@link IMultipleTankHandler#allowSameFluidFill()}.
+     */
+    final class MultiFluidTankEntry implements IFluidTank {
+
+        private final IMultipleTankHandler tank;
+        private final IFluidTank delegate;
+
+        public MultiFluidTankEntry(@Nonnull IMultipleTankHandler tank, @Nonnull IFluidTank delegate) {
+            this.tank = tank;
+            this.delegate = delegate;
+        }
+
+        @Nonnull
+        public IMultipleTankHandler getTank() {
+            return tank;
+        }
+
+        @Nonnull
+        public IFluidTank getDelegate() {
+            return delegate;
+        }
+
+        public boolean allowSameFluidFill() {
+            return tank.allowSameFluidFill();
+        }
+
+        @Nonnull
+        public IFluidTankProperties[] getTankProperties() {
+            return delegate instanceof IFluidHandler ?
+                    ((IFluidHandler) delegate).getTankProperties() :
+                    new IFluidTankProperties[]{new FallbackTankProperty()};
+        }
+
+        @SuppressWarnings("unchecked")
+        public NBTTagCompound trySerialize() {
+            if (delegate instanceof FluidTank) {
+                return ((FluidTank) delegate).writeToNBT(new NBTTagCompound());
+            } else if (delegate instanceof INBTSerializable) {
+                try {
+                    return ((INBTSerializable<NBTTagCompound>) delegate).serializeNBT();
+                } catch (ClassCastException ignored) {}
```

Comment:
Is this try-catch block needed?

Reply:
`INBTSerializable` implementations can return other types of `NBTBase`, so the try-catch accounts for that.

Reply:
How frequently will this catch an exception? If it's infrequent, should a log with the error be made here?

Reply:
I think ignoring the exception is better than logging things, since logs can get quite spammy. Also I don't see how I should write the log with actually helpful information that isn't just a bunch of stacktrace and toString of the delegate instance.

Fluid tanks using different NBT type just means it's incompatible with save/load functionality of FluidTankList. If we're logging save/load failure for incompatible `INBTSerializable`s, we might as well log the cases where the delegate doesn't implement `INBTSerializable` from the start.